juice/src/net/common/convolution.rs

```diff
@@ -0,0 +1,238 @@
+        // convolution context for it. Store it in the context cache to reuse later.
+        let mut convolution_context_ref = self.convolution_contexts.borrow_mut();
+        let mut convolution_context = convolution_context_ref.entry(context.batch_size()).or_insert_with(|| {
+            backend
+                .new_convolution_context(
+                    &input.borrow(),
+                    &output.borrow(),
+                    &self.kernel.borrow().data,
+                    conn::ConvForwardAlgo::Auto,
+                    conn::ConvBackwardFilterAlgo::Auto,
+                    conn::ConvBackwardDataAlgo::Auto,
+                    &[self.config.stride as i32, self.config.stride as i32],
+                    &[self.config.padding as i32, self.config.padding as i32],
+                )
+                .unwrap()
+        });
+
+        backend
+            .convolution(
+                &self.kernel.borrow().data,
+                &input.borrow(),
+                &mut output.borrow_mut(),
+                &mut convolution_context,
+            )
+            .unwrap();
+    }
+
+    fn compute_gradients(&self, backend: &B, context: &mut Context) {
+        let input = context.get_data(self.descriptor.input(0));
+        let output_gradient = context.get_data_gradient(self.descriptor.output(0));
+
+        let input_gradient = context.acquire_data_gradient(self.descriptor.input(0));
+        let kernel_gradient = context.acquire_params_gradient(self.descriptor.param(0));
+
+        // At this point the convolution context must already exist in the contexts cache
+        // (should have been created in the compute_output() function),
+        // so we just take it from the cache.
+        let mut convolution_context_ref = self.convolution_contexts.borrow_mut();
+        let mut convolution_context = convolution_context_ref.get_mut(&context.batch_size()).unwrap();
+
+        // Network error gradient with respect to input data.
+        backend
+            .convolution_grad_data(
+                &self.kernel.borrow().data,
+                &output_gradient.borrow(),
+                &mut input_gradient.borrow_mut(),
+                convolution_context,
+            )
+            .unwrap();
+
+        // Network error gradient with respect to the kernel.
+        backend
+            .convolution_grad_filter(
+                &input.borrow(),
+                &mut output_gradient.borrow(),
+                &mut kernel_gradient.borrow_mut(),
+                &mut convolution_context,
+            )
+            .unwrap();
+    }
+
+    fn descriptor(&self) -> &Descriptor {
+        &self.descriptor
+    }
+
+    fn descriptor_mut(&mut self) -> &mut Descriptor {
+        &mut self.descriptor
+    }
+}
+
+impl<B: conn::Convolution<f32>> Debug for Convolution<B> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Convolution")
```

Comment:
Should the dimensionality of the weights be included?

Reply:
Done.